File: DSA/Priyanshi/Day13_que1.java
import java.util.*;

public class Day13_que1 {
    public static String wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true; 
        
        for (int j = 1; j <= n; j++) {
            for (int i = 0; i < j; i++) {
                if (dp[i] && wordSet.contains(s.substring(i, j))) {
                    dp[j] = true;
                    break;
                }
            }
        }
        
        return dp[n] ? "YES" : "NO";
    }

    public static void main(String[] args) {
        String s = "applepie";
        List<String> dictionary = Arrays.asList("apple", "pie");

        String result = wordBreak(s, dictionary);
        System.out.println("Can the string be segmented? " + result);
    }
}

File: DSA/Devansh-Nishchal-Mandloi/day13que1.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <string>

using namespace std;

string wordBreak(string s, vector<string>& wordDict) {
    unordered_set<string> dict(wordDict.begin(), wordDict.end());
    vector<bool> dp(s.length() + 1, false);
    dp[0] = true;  

    for (int i = 1; i <= s.length(); ++i) {
        for (int j = 0; j < i; ++j) {
            if (dp[j] && dict.find(s.substr(j, i - j)) != dict.end()) {
                dp[i] = true;
                break;
            }
        }
    }

    return dp[s.length()] ? "YES" : "NO";
}

int main() {
    string s = "applepie";
    vector<string> wordDict = {"apple", "pie"};

    cout << "Output: " << wordBreak(s, wordDict) << endl; 

    return 0;
}
